File: grokking-the-coding-interview_patterns-for-coding-questions/1.-pattern-two-pointers/6-triplets-with-smaller-sum-medium/solution_6-triplets-with-smaller-sum-medium.py
def three_sum_smaller(nums, target): 
    nums.sort()

    count = []

    for i in range(len(nums) - 1):
        l = i + 1
        r = len(nums) - 1
        curr_sum = nums[i] + nums[l] + nums[r]

        while l < r:
            if curr_sum < target:
                count.append([nums[i], nums[l], nums[r]])
                l += 1
            else:
                r -= 1

    return len(count)


print(three_sum_smaller([-2, 0, 1, 3], 2)) #2
print(three_sum_smaller([-2, 0, -1, 3], 2)) #2
print(three_sum_smaller([3, 1, 0, 2], 5)) #2

File: grokking-the-coding-interview_patterns-for-coding-questions/1.-pattern-two-pointers/6-triplets-with-smaller-sum-medium/solution_6-triplets-with-smaller-sum-medium.js
const tripletSmallerSum = (nums, target) => {
    nums.sort((a, b) => a-b);
    let count = 0;
    for (let i = 0; i < nums.length - 2; i++) {
        let l = i + 1;
        let r = nums.length - 1;
        let current = nums[i] + nums[l] + nums[r];
        while (l < r) {
            if (current < target) {
                count++;
                l++;
            } else {
                r--;
            }
        }
    }
    return count;

}

console.log(tripletSmallerSum([-2,0,1,3], 2));
console.log(tripletSmallerSum([-2,0,-1,3], 2));
console.log(tripletSmallerSum([3, 1, 0, 2], 5));